README.md
已适配 null-safe ，对应版本：2.0.0

## 导语：

对于任何一款应用而言，页面的流畅度一定是影响用户体验最重要的几个指标之一。作为开发者，优化页面流畅度也是自己技术实力的体现。但在决定进行优化之前，还有两个更重要的问题摆在我们面前：**1、如何发现卡顿的页面？**、**2、如何衡量我的优化效果？**

为了解决这两个问题，本期给大家带来一个很有意思的小工具：**fps_monitor**

***
## 一、What's this 这是个什么工具？

首先一句话告诉你：**这是一个能在 profile/debug 模式下，直观帮助我们评估页面流畅度的工具！！**
直白来说就是：这是一个可以在(刷新率60)设备上直接查看最近（默认 100）帧的表现情况的小工具，直接上图：



当我们点击右下角的 ⏯ 后，按钮变为⏸ 状态，工具开始为我们收集每一帧的总耗时(包含 CPU 和 GPU 耗时)。此时点击 ⏸  按钮会为我们展示收集到的耗时信息

![image.png](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/a36915b137e14d1e9d4e9bbbe7f5bd6c~tplv-k3u1fbpfcp-watermark.image)

柱状图顶部为我们展示收集到的数据里：**最大耗时**、**平均耗时**、以及**总耗时**（单位：毫秒）

在下方，我将页面流畅度划为了四个级别：**流畅（蓝色）**、**良好(黄色)**、**轻微卡顿(粉色)**、**卡顿(红色)**，将 FPS 折算成一帧所消耗的时间，不同级别采用不一样的颜色，统计不同级别出现的次数。

上面的例子中，我们可以看到，工具一共收集了 99 帧，最大耗时的一帧花了119ms，平均耗时：32.9 ms，总耗时：3259.5 ms。其中认为有 40 帧流畅，25 帧良好，38 帧轻微卡顿，6 帧卡顿。



***

## 二、Why you need this 为什么你需要这个工具？

### 1、为什么我没选择 PerformanceOverLay 和 DoKit？
看到上面的功能可能有人有疑惑，你这功能咋和 PerformanceOverLay 这么类似？

首先，我在使用 PerformanceOverLay 的时候遇到了一点问题：


![image.png](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/151f8e7f24ae4f09a3583c7e173bbf6c~tplv-k3u1fbpfcp-watermark.image)

如图，PerformanceOverLay 上分别为我们展示了构建（UI）耗时和渲染（GPU）耗时。

我遇到的第一个问题是，因为我们在判断流畅度的时候，**往往是看一帧的总耗时**。这样拆分之后，一帧的耗时变成了上下的和，对我而言很不直观。

其次，这里面提供`最大耗时`或者`平均耗时`并不能很好的帮助我们量化页面的流畅程度。因为这个统计过程，会直接将一帧的耗时进行平均，这就带来一个问题。我们知道对于60刷新率的设备，两帧的间隔时间最小应该是 16.7ms，而 PerformanceOverLay 的收集过程没有对数据过滤，会出现一帧耗时小于 16.7ms，这就导致平均数据可能偏低。（下图平均一帧耗时为：10.6ms 60HZ设备）
![image.png](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/3289936d52424f7797f43634837148ce~tplv-k3u1fbpfcp-watermark.image)

其实这样来看，DoKit是一个不错的选择

![image.png](https://p1-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/57838121f69d4694b7ee074bbb7d1e19~tplv-k3u1fbpfcp-watermark.image)

但DoKit同样没有对最小帧耗时做过滤，也会出现平均耗时偏低的情况。同时，没有更多的数据辅助评估页面的流畅程度。

上面我遇到的情况，不一定是问题，只是我在使用过程中觉得不太直观，不太方便。

因此开发了这个工具，该工具具有以下特点


![image.png](https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/fde2765e3e944dbf98e90a2ffea75fad~tplv-k3u1fbpfcp-watermark.image)

同时支持设置最大采集帧数

### 2、我是如何理解页面流畅度

我在上一期[ListView流畅度翻倍！！Flutter卡顿分析和通用优化方案](https://juejin.cn/post/6940134891606507534)有解释过

对于大部分人而言，当每秒的画面达到60，也就是俗称**60FPS**的时候，整个过程就是流畅的。

![](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/b309e9f3ffe8425abb5ba11f5ef1f669~tplv-k3u1fbpfcp-zoom-1.image)

一秒 60 帧，也就意味着平均两帧之间的间隔为 16.7ms。那么耗时大于 16.7ms 就会觉得卡顿么？

答案当然是 NO。腾讯在 [Martrix](https://github.com/Tencent/matrix/wiki/Matrix-Android-TraceCanary) 中也提到

>我们平时看到的大部分电影或视频 FPS 其实不高，一般只有 25FPS ~ 30FPS，而实际上我们也没有觉得卡顿。 在人眼结构上看，当一组动作在 1 秒内有 12 次变化（即 12FPS），我们会认为这组动作是连贯的

其实流畅度本身就是一个很主观的东西，就好比有人觉得打王者荣耀不开高帧率好像也还算流畅，有人觉得不开高帧率那不就是个GIF图么。

有没有客观一点的指标，我在网上查询了很久之后找到了一篇08年发表在ICIP上的论文[Modeling the impact of frame rate on perceptual quality of video
](https://www.researchgate.net/profile/Zhan-Ma-6/publication/224359119_Modeling_the_impact_of_frame_rate_on_perceptual_quality_of_video/links/00b7d514f3b347250e000000/Modeling-the-impact-of-frame-rate-on-perceptual-quality-of-video.pdf) 他们使用了6种内容进行测试，实验结果如下图所示

![image.png](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/3dbfc676b92e48ecbcddfb08d9da8a6e~tplv-k3u1fbpfcp-watermark.image)

通过该图我们可以看出，当帧率大于15帧的时候，人眼的主观感受差别不大，基本上都处于较高的水平。而帧率小于15帧以后，人眼的主观感受会急剧下降。换句话说，人眼会立刻感受到画面的不连贯性。


因此，在工具中我将低于16.7ms的数据统一成16.7ms,所以这个检测工具只在刷新率为60的设备有意义。并且将流畅度划分为了以下等级：

* 流畅：FPS大于55，即一帧耗时低于 18ms
* 良好：FPS在30-55之间，即一帧耗时在 18ms-33ms 之间
* 轻微卡顿：FPS在15-30之间，即一帧耗时在 33ms-67ms 之间
* 卡顿：FPS低于15，即一帧耗时大于 66.7ms

并统计出现的次数，你可以根据这几项数据，对比优化前后的数据，得出性能的提升情况；当然也可以制定一个理想的流畅度。例如：流畅的帧数占统计帧数的90%，或者卡顿的帧数不超过5次。

***

##  三、How to use it 如何使用？
### 1、项目依赖
dependencies:
 fps_monitor: ^1.12.13-1
### 2、接入工程
有两处接入点

* 指定overLayState ，因为需要弹出一个Fps的统计页面，所以当前指定overLayState。
**（PS：大家一般使用Navigator.of(context)去跳转一个页面，通过GlobalKey可以实现无context的跳转）**

![image.png](https://p1-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/800dd3bb2e4c402db6d66758ad6154e2~tplv-k3u1fbpfcp-watermark.image)
```dart
///声明NavigatorState的GlobalKey
GlobalKey<NavigatorState> globalKey = GlobalKey();
///获取overLayState
  SchedulerBinding.instance.addPostFrameCallback((t) =>
    overlayState =globalKey.currentState.overlay
    );
///指定MaterialApp的navigatorKey  
navigatorKey: globalKey,

```

* 在build属性中包裹组件

![image.png](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/5770b6ef7ede429c95176e1eceb858b6~tplv-k3u1fbpfcp-watermark.image)

```dart
builder: (ctx, child) =>
          CustomWidgetInspector(
            child: child,
          ),
```

### 3、如何使用
在完成了上述步骤之后，你只需要启动app，**该工具只会在profile/debug模式下集成**，在你的右下角会出现一个 ⏯ 按钮，点击开始记录，再次点击显示数据。

![Screenrecording_20210405_171133.gif](https://p1-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/4b13856e7c274513ac488f2959877786~tplv-k3u1fbpfcp-watermark.image)

如果想要结束采集，点击面板中的停止监听即可。

如果你想采集更多的帧，可以通过`kFpsInfoMaxSize`设置

![Screenrecording_20210405_171154.gif](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/373a7a22b17d4b308eacaaaa933a3358~tplv-k3u1fbpfcp-watermark.image)

### 4、Warning：

空安全之前 请使用 1.12.13-2 分支
空安全后 使用 2.0.0 分支

***

### 四、how do it 来点原理？

可能你会对这个工具的检测原理感兴趣，那咱们再来唠两句原理。

#### 绘制数据的获取
```dart
 WidgetsBinding.instance.addTimingsCallback(monitor);
```
Flutter 会在每帧完成绘制后，会将耗时进行回调。耗时体现在三个变量上：1、构建时间；2、绘制时间；3、总时间。当你点击 ⏯ 按钮的时候，工具便开始采集耗时信息。

其实对于 Flutter 相关的渲染调度，推荐大家看看 `SchedulerBinding` ，里面写得再详细不过。

#### 显示Fps界面

显示 Fps 的页面比较简单，直接通过 OverlayState 插入即可。如果你不太熟悉 Overlay 可以把它理解成浮窗。其中的表格绘制通过使用 DoKit 的自定义画笔实现，当然也可替换成各种开源的图表库。

***
## 最后 感谢各位吴彦祖和彭于晏的点赞,Start,和Follow

如果你觉得这个工具还不错，点个赞支持一下吧~

最后附上我的：

掘金主页：[Nayuta]( https://juejin.cn/user/4309694831660711)


欢迎搜索公众号：**进击的Flutter或者runflutter** 里面整理收集了最详细的 Flutter 进阶与优化指南。关注我，获取我的最新文章~


![1361617636891_.pic.jpg](https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/dd8c5b194fe244d7a8b56a68e254a4fb~tplv-k3u1fbpfcp-watermark.image)


如果使用过程有问题可以直接通过公众号私信我~


CHANGELOG.md
## [1.12.13-0] -正式稳定版本
## [1.12.13-1] -更新readme
## [1.12.13-2] -更新readme
## [2.0.0] -适配空安全


test/fps_monitor_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:fps_monitor/fps_monitor.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'package:fps_monitor/widget/custom_widget_inspector.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    GlobalKey<NavigatorState> globalKey = GlobalKey();
    WidgetsBinding.instance.addPostFrameCallback((t) {
      overlayState = globalKey.currentState.overlay;
    });
    return MaterialApp(
      navigatorKey: globalKey,
      title: 'Flutter Demo',
      builder: (ctx, child) => CustomWidgetInspector(
        child: child,
      ),
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
            FloatingActionButton(
              onPressed: _incrementCounter,
              tooltip: 'Increment',
              child: Icon(Icons.add),
            )
          ],
        ),
      ),
    );
  }
}


lib/widget/fps_page.dart
import 'package:flutter/material.dart';
import 'package:fps_monitor/bean/fps_info.dart';
import 'package:fps_monitor/util/collection_util.dart';

import 'fps_chart.dart';

class FpsPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return FpsPageState();
  }
}

class FpsPageState extends State<FpsPage> {
  @override
  Widget build(BuildContext context) {
    List<FpsInfo> list = CommonStorage.instance!.getAll();
    return Container(
      width: MediaQuery.of(context).size.width,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.start,
        children: <Widget>[
          Container(
              height: 44,
              padding: const EdgeInsets.only(left: 20, right: 20),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: [
                  info('近${CommonStorage.instance!.items.length}帧'),
                  info('最大耗时:${CommonStorage.instance!.max!.toStringAsFixed(1)}'),
                  info(
                      '平均耗时:${CommonStorage.instance!.getAvg().toStringAsFixed(1)}'),
                  info(
                      '总耗时:${CommonStorage.instance!.totalNum.toStringAsFixed(1)}')
                ],
              )),
          Divider(
            height: 0.5,
            color: Color(0xffdddddd),
            indent: 16,
            endIndent: 16,
          ),
          FpsBarChart(data: list)
        ],
      ),
    );
  }

  Widget info(String text) {
    return Text("$text",
        style: const TextStyle(
            color: Colors.grey,
            fontWeight: FontWeight.normal,
            fontFamily: 'PingFang SC',
            fontSize: 9));
  }
}


lib/util/collection_util.dart
import 'dart:collection';

import 'package:fps_monitor/bean/fps_info.dart';

int kFpsInfoMaxSize = 100;

/// Fps 信息存储
class CommonStorage {
  static CommonStorage? _instance;

  static CommonStorage? get instance {
    if (_instance == null) {
      _instance = CommonStorage._();
    }
    return _instance;
  }

  CommonStorage._();

  int maxCount = kFpsInfoMaxSize;
  Queue<FpsInfo> items = new Queue();
  double? max = 0;
  double totalNum = 0;

  List<FpsInfo> getAll() {
    return items.toList();
  }

  void clear() {
    items.clear();
    max = 0;
    totalNum = 0;
  }

  bool save(FpsInfo info) {
    if (items.length >= maxCount) {
      totalNum -= items.removeFirst().totalSpan!;
    }
    items.add(info);
    totalNum += info.totalSpan!;
    max = info.totalSpan! > max! ? info.totalSpan : max;
    return true;
  }

  num getAvg() {
    return totalNum / items.length;
  }

  bool contains(FpsInfo info) {
    return items.contains(info);
  }
}


lib/widget/custom_widget_inspector.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';

import 'performance_observer_widget.dart';

const double _kInspectButtonMargin = 10.0;
const double _kErrorReminderButtonMargin = 40.0;
late OverlayState overlayState;

class CustomWidgetInspector extends StatefulWidget {
  /// 展示性能监控数据
  static bool debugShowPerformanceMonitor = true;

  /// Creates a widget that enables inspection for the child.
  ///
  /// The [child] argument must not be null.
  const CustomWidgetInspector({Key? key, required this.child})
      : assert(child != null),
        super(key: key);

  /// The widget that is being inspected.
  final Widget child;

  @override
  _CustomWidgetInspectorState createState() => _CustomWidgetInspectorState();
}

class _CustomWidgetInspectorState extends State<CustomWidgetInspector> {
  /// Distance from the edge of the bounding box for an element to consider
  /// as selecting the edge of the bounding box.

  final GlobalKey rootGlobalKey = GlobalKey();
  final InspectorSelection selection =
      WidgetInspectorService.instance.selection;

  @override
  Widget build(BuildContext context) {
    Widget child = widget.child;

    //release模式不打开FPS监控
    if (bool.fromEnvironment("dart.vm.product")) return child;

    Widget performanceObserver = Container();

    if (CustomWidgetInspector.debugShowPerformanceMonitor) {
      performanceObserver = PerformanceObserverWidget();
    }

    if (CustomWidgetInspector.debugShowPerformanceMonitor) {
      child = Stack(
        alignment: AlignmentDirectional.topStart,
        children: <Widget>[
          IgnorePointer(
            ignoring: false,
            key: rootGlobalKey,
            child: child,
          ),
          Positioned(
            right: _kInspectButtonMargin,
            bottom: _kErrorReminderButtonMargin,
            child: performanceObserver,
          )
        ],
      );
    }
    return child;
  }

  @override
  void dispose() {
    super.dispose();
  }
}


lib/widget/performance_observer_widget.dart
import 'dart:math';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:fps_monitor/bean/fps_info.dart';
import 'package:fps_monitor/util/collection_util.dart';
import 'package:fps_monitor/widget/custom_widget_inspector.dart';

import 'fps_page.dart';

class PerformanceObserverWidget extends StatefulWidget {
  const PerformanceObserverWidget({Key? key}) : super(key: key);

  @override
  _PerformanceObserverWidgetState createState() =>
      _PerformanceObserverWidgetState();
}

class _PerformanceObserverWidgetState extends State<PerformanceObserverWidget> {
  bool startRecording = false;
  bool fpsPageShowing = false;

  late ValueNotifier controller;
  late Function(List<FrameTiming>) monitor;
  OverlayEntry? fpsInfoPage;
  OverlayEntry? performancePage;

  @override
  void initState() {
    super.initState();
    controller = ValueNotifier("");
    monitor = (timings) {
      double duration = 0;
      timings.forEach((element) {
        FrameTiming frameTiming = element;
        duration = frameTiming.totalSpan.inMilliseconds.toDouble();
        FpsInfo fpsInfo = new FpsInfo();
        fpsInfo.totalSpan = max(16.7, duration);
        CommonStorage.instance!.save(fpsInfo);
      });
    };
  }

  @override
  void dispose() {
    stop();
    super.dispose();
  }

  void start() {
    WidgetsBinding.instance!.addTimingsCallback(monitor);
  }

  void stop() {
    WidgetsBinding.instance!.removeTimingsCallback(monitor);
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        GestureDetector(
          child: RepaintBoundary(
            child: ValueListenableBuilder(
                valueListenable: controller,
                builder: (context, dynamic snapshot, _) {
                  return Container(
                    color: Colors.white,
                    child: !startRecording
                        ? Icon(Icons.play_arrow)
                        : fpsPageShowing ? Row() : Icon(Icons.pause),
                  );
                }),
          ),
          onTap: () {
            fpsMonitor();
          },
        )
      ],
    );
  }

  void fpsMonitor() {
    if (!startRecording) {
      setState(() {
        start();
        startRecording = true;
        controller.value = startRecording;
      });
    } else {
      if (!fpsPageShowing) {
        stop();
        if (fpsInfoPage == null) {
          fpsInfoPage = OverlayEntry(builder: (c) {
            return Scaffold(
              body: Column(
                children: <Widget>[
                  Expanded(
                      child: GestureDetector(
                    onTap: () {
                      fpsInfoPage!.remove();
                      fpsPageShowing = false;
                      start();
                    },
                    child: Container(
                      color: Color(0x33999999),
                    ),
                  )),
                  Container(
                      color: Colors.white,
                      child: Column(
                        children: <Widget>[
                          FpsPage(),
                          Divider(),
                          Container(
                            padding: const EdgeInsets.only(
                                left: 20, top: 20, bottom: 20),
                            child: GestureDetector(
                              child: Text(
                                '停止监听',
                                style: TextStyle(color: Colors.blue),
                              ),
                              onTap: () {
                                startRecording = false;
                                fpsInfoPage!.remove();
                                fpsPageShowing = false;
                                CommonStorage.instance!.clear();
                                controller.value = startRecording;
                                // setState(() {});
                              },
                            ),
                            alignment: Alignment.bottomLeft,
                          ),
                        ],
                      )),
                ],
              ),
              backgroundColor: Color(0x33999999),
            );
          });
        }
        fpsPageShowing = true;
        overlayState.insert(fpsInfoPage!);
      }
    }
  }
}


lib/widget/fps_chart.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:fps_monitor/bean/fps_info.dart';
import 'package:fps_monitor/util/collection_util.dart';

class BarChartPainter extends CustomPainter {
  List<FpsInfo> datas;

  BarChartPainter({required this.datas});

  @override
  bool shouldRepaint(BarChartPainter oldDelegate) => true;

  @override
  bool shouldRebuildSemantics(BarChartPainter oldDelegate) => false;

  void _drawAxis(Canvas canvas, Size size) {
    final double sw = size.width;
    final double sh = size.height;

    // 使用 Paint 定义路径的样式
    final Paint paint = Paint()
      ..color = Color(0xffdddddd)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 0.5;

    // 使用 Path 定义绘制的路径，从画布的左上角到左下角在到右下角
    final Path path = Path()
      ..moveTo(0, 0)
      ..lineTo(0, sh)
      ..lineTo(sw, sh);

    // 使用 drawPath 方法绘制路径
    canvas.drawPath(path, paint);
  }

  void _drawLabels(Canvas canvas, Size size) {
    double labelFontSize = 10;
    final double sh = size.height;
    final List<double> yAxisLabels = [];

    yAxisLabels.add(16);
    yAxisLabels.add(33);
    yAxisLabels.add(66);
    yAxisLabels.add(100);

    yAxisLabels.asMap().forEach(
      (index, label) {
        // 标识的高度为画布高度减去标识的值
        final double top = sh - label * 2.5;
        // final rect = Rect.fromLTWH(0, top, 4, 1);
        final Offset textOffset = Offset(
          0 - (label.toInt().toString().length == 3 ? 24 : 20).toDouble(),
          top - labelFontSize / 2,
        );

        // 绘制文字需要用 `TextPainter`，最后调用 paint 方法绘制文字
        TextPainter(
          text: TextSpan(
            text: label.toStringAsFixed(0),
            style: TextStyle(fontSize: labelFontSize, color: Color(0xff4a4b5b)),
          ),
          textAlign: TextAlign.right,
          textDirection: TextDirection.ltr,
          textWidthBasis: TextWidthBasis.longestLine,
        )
          ..layout(minWidth: 0, maxWidth: 24)
          ..paint(canvas, textOffset);
      },
    );
  }

  void _drawBars(Canvas canvas, Size size) {
    final sh = size.height;
    final paint = Paint()..style = PaintingStyle.fill;
    final double marginLeft = 7.5;
    double _barWidth = (size.width / CommonStorage.instance!.maxCount);
    double _barGap = 0;
    int A = 0;
    int B = 0;
    int C = 0;
    int D = 0;

    for (int i = 0; i < datas.length; i++) {
      int value = datas[i].getValue()!.toInt();

      if (value > 66) {
        D++;
      } else if (value > 33) {
        C++;
      } else if (value > 18) {
        B++;
      } else {
        A++;
      }
      paint.color = value <= 18
          ? Color(0xff55a8fd)
          : value <= 33
              ? Color(0xfffad337)
              : value <= 66 ? Color(0xFFF48FB1) : Color(0xFFD32F2F);
      // 矩形的上边缘为画布高度减去数据值
      final double top = sh - value * 2.5;
      // 矩形的左边缘为当前索引值乘以矩形宽度加上矩形之间的间距
      final double left = marginLeft + i * _barWidth + (i * _barGap) + _barGap;

      // 使用 Rect.fromLTWH 方法创建要绘制的矩形
      final rect = Rect.fromLTWH(left, top, _barWidth, value * 2.5.toDouble());
      // 使用 drawRect 方法绘制矩形
      canvas.drawRect(rect, paint);
    }
    TextPainter(
      text: TextSpan(
        text: "流畅：$A 良好：$B 轻微卡顿：$C 卡顿：$D",
        style: TextStyle(fontSize: 10, color: Color(0xff4a4b5b)),
      ),
      textAlign: TextAlign.right,
      textDirection: TextDirection.ltr,
      textWidthBasis: TextWidthBasis.longestLine,
    )
      ..layout(
        minWidth: 0,
      )
      ..paint(canvas, Offset.zero);
  }

  @override
  void paint(Canvas canvas, Size size) {
    _drawAxis(canvas, size);
    _drawLabels(canvas, size);
    _drawBars(canvas, size);
  }
}

class FpsBarChart extends StatefulWidget {
  final List<FpsInfo> data;

  const FpsBarChart({
    required this.data,
  });

  @override
  _FpsBarChartState createState() => _FpsBarChartState();
}

class _FpsBarChartState extends State<FpsBarChart>
    with TickerProviderStateMixin {
  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).orientation == Orientation.portrait
        ? MediaQuery.of(context).size.width - 56
        : MediaQuery.of(context).size.width - 30;
    double height = MediaQuery.of(context).orientation == Orientation.portrait
        ? MediaQuery.of(context).size.height - 500
        : MediaQuery.of(context).size.height - 200;
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Container(
          margin: EdgeInsets.only(top: 40, left: 24),
          child: CustomPaint(
            painter: BarChartPainter(datas: widget.data),
            child: Container(
              width: width,
              height: height,
            ),
          ),
        )
      ],
    );
  }
}
